channeldb/reports.go

```diff
@@ -0,0 +1,288 @@
+package channeldb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+var (
+	// reportBucket is a bucket which holds a record of the final set
+	// of resolutions which were processed for a channel.
+	// [reportBucket]
+	//	[chainHashBucket]
+	//		[channelBucket]
+	//			[resolversBucket]
+	reportBucket = []byte("channel-reports")
+
+	// resolversBucket holds the outcome of a channel's resolvers.
+	resolversBucket = []byte("resolvers-bucket")
+)
+
+var (
+	// ErrNoReportBucket is returned if the report bucket has not been
+	// created yet.
+	ErrNoReportBucket = errors.New("no reports bucket present")
+
+	// ErrNoChainHashBucket is returned when we have not created a bucket
+	// for the current chain hash.
+	ErrNoChainHashBucket = errors.New("no chain hash bucket")
+
+	// ErrNoChannel is returned when a channel is not found in the chain
+	// hash bucket.
+	ErrNoChannel = errors.New("channel bucket not found")
+)
+
+// ResolverOutcome describes the outcome for the resolver that the contract
+// court reached. This state is not necessarily final, htlcs on our own
+// commitment are resolved across two resolvers.
+type ResolverOutcome uint8
+
+const (
+	resolverOutcomeUnknown ResolverOutcome = iota
+)
+
+// ResolverReport provides an account of the outcome of a resolver. This differs
+// from a ContractReport because it does not necessarily fully resolve the
+// contract; each step of two stage htlc resolution is included.
+type ResolverReport struct {
+	// OutPoint is the outpoint that was spent by this resolver.
+	OutPoint wire.OutPoint
+
+	// Amount is the amount in satoshis that was claimed by this resolution.
+	Amount btcutil.Amount
+
+	// ResolverOutcome indicates the type of resolution that occurred.
+	ResolverOutcome
+
+	// SpendTxID is the transaction ID of the spending transaction that
+	// claimed the outpoint. This may be a sweep transaction, or a first
+	// stage success/timeout transaction.
+	SpendTxID *chainhash.Hash
+}
+
+// PutResolverReport creates and commits a transaction that is used to write a
+// resolver report to disk.
+func (d *DB) PutResolverReport(chainHash chainhash.Hash,
+	channelOutpoint *wire.OutPoint, report *ResolverReport) error {
+
+	tx, err := d.BeginReadWriteTx()
+	if err != nil {
+		return err
+	}
+
+	if err := PutResolverReportInTx(
+		tx, chainHash, channelOutpoint, report,
+	); err != nil {
+		if err := tx.Rollback(); err != nil {
+			return err
+		}
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// PutResolverReportInTx writes a resolver report to its channel's bucket within
+// a preexisting transaction.
+func PutResolverReportInTx(tx kvdb.RwTx, chainHash chainhash.Hash,
```

Comment:
To cut down on the redundant amount of methods a bit, one pattern we follow elsewhere is to make a new transaction if the passed transaction is `nil`.